### PR TITLE
chore(broker): split snapshot replication destination folders

### LINKED ISSUE #3531: Failed to delete pending snapshot
**Describe the bug**
It seems that on role transition we sometimes are not able to delete pending snapshots, might be a race condition. Idk how bad it is.

**Expected behavior**
No exception.

**Log/Stacktrace**
<details><summary>Stacktrace</summary>
 <p>

```
I 2019-12-12T00:51:48.587970665Z 2019-12-12 00:51:48.584 [Broker-4-TopologyManager] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-0] DEBUG io.zeebe.broker.clustering - Received metadata change for 2, partitions {1=FOLLOWER, 2=FOLLOWER, 3=FOLLOWER, 6=FOLLOWER, 7=FOLLOWER, 8=FOLLOWER} terms {}
 
I 2019-12-12T00:51:50.054779659Z 2019-12-12 00:51:50.054 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-8}{role=FOLLOWER} - Accepted PollRequest{term=1, candidate=2, lastLogIndex=52383006, lastLogTerm=1}: candidate's log is up-to-date
 
I 2019-12-12T00:51:50.070174843Z 2019-12-12 00:51:50.069 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-8}{role=FOLLOWER} - Accepted VoteRequest{term=2, candidate=2, lastLogIndex=52383006, lastLogTerm=1}: candidate's log is up-to-date
 
I 2019-12-12T00:51:53.090583234Z 2019-12-12 00:51:53.090 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.roles.FollowerRole - RaftServer{raft-partition-partition-8}{role=FOLLOWER} - No heartbeat from null in the last PT3.018S (calculated from last 3018 ms)
 
I 2019-12-12T00:51:53.095958119Z 2019-12-12 00:51:53.095 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-8} - Transitioning to CANDIDATE
 
I 2019-12-12T00:51:53.097347856Z 2019-12-12 00:51:53.096 [] [raft-server-4-raft-partition-partition-8] WARN  io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot - Failed to delete pending snapshot DbPendingSnapshot{index=22428805, term=1, timestamp=2019-12-11 04:07:45,487, directory=/usr/local/zeebe/data/raft-partition/partitions/8/pending/22428805-1-1576080465487}
 
I 2019-12-12T00:51:53.097377276Z java.nio.file.NoSuchFileException: /usr/local/zeebe/data/raft-partition/partitions/8/pending/22428805-1-1576080465487
	at sun.nio.fs.UnixException.translateToIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(Unknown Source) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.readAttributes(Unknown Source) ~[?:?]
	at sun.nio.fs.LinuxFileSystemProvider.readAttributes(Unknown Source) ~[?:?]
	at java.nio.file.Files.readAttributes(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.getAttributes(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.visit(Unknown Source) ~[?:?]
	at java.nio.file.FileTreeWalker.walk(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at java.nio.file.Files.walkFileTree(Unknown Source) ~[?:?]
	at io.zeebe.util.FileUtil.deleteFolder(FileUtil.java:41) ~[zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.broker.clustering.atomix.storage.snapshot.DbPendingSnapshot.abort(DbPendingSnapshot.java:120) ~[zeebe-broker-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.atomix.protocols.raft.roles.PassiveRole.stop(PassiveRole.java:83) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.roles.FollowerRole.stop(FollowerRole.java:88) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.impl.RaftContext.transition(RaftContext.java:532) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.roles.FollowerRole.lambda$sendPollRequests$3(FollowerRole.java:152) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.utils.Quorum.checkComplete(Quorum.java:48) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.utils.Quorum.succeed(Quorum.java:40) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at io.atomix.protocols.raft.roles.FollowerRole.lambda$sendPollRequests$4(FollowerRole.java:204) ~[atomix-raft-3.2.0-20191207.093106-100.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$1.lambda$execute$0(SingleThreadContext.java:53) ~[atomix-utils-3.2.0-20191207.093016-100.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
 
I 2019-12-12T00:51:53.097948519Z 2019-12-12 00:51:53.097 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.roles.CandidateRole - RaftServer{raft-partition-partition-8}{role=CANDIDATE} - Starting election
 
I 2019-12-12T00:51:53.103847431Z 2019-12-12 00:51:53.103 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-3] DEBUG io.zeebe.broker.system - Partition role transitioning from FOLLOWER to CANDIDATE
 
I 2019-12-12T00:51:53.112471577Z 2019-12-12 00:51:53.112 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-8} - Transitioning to LEADER
 
I 2019-12-12T00:51:53.112908081Z 2019-12-12 00:51:53.112 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{raft-partition-partition-8} - Found leader 4
 
I 2019-12-12T00:51:59.760709340Z 2019-12-12 00:51:59.760 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.broker.system - Removing follower partition service for partition PartitionId{id=8, group=raft-partition}
 
I 2019-12-12T00:51:59.760743317Z 2019-12-12 00:51:59.760 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.broker.system - Closing io.zeebe.broker.logstreams.LogStreamDeletionService
 
I 2019-12-12T00:51:59.760750238Z 2019-12-12 00:51:59.760 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.broker.system - Partition role transitioning from CANDIDATE to LEADER
 
I 2019-12-12T00:51:59.760819365Z 2019-12-12 00:51:59.760 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.logstreams.LogStreamDeletionService successfully
 
I 2019-12-12T00:51:59.951003303Z 2019-12-12 00:51:59.950 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Closed database from '/usr/local/zeebe/data/raft-partition/partitions/8/runtime'.
 
I 2019-12-12T00:51:59.951057977Z 2019-12-12 00:51:59.950 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.broker.system - Installing leader partition service for partition PartitionId{id=8, group=raft-partition}
 
I 2019-12-12T00:52:00.019697783Z 2019-12-12 00:52:00.019 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=126104534841528, path=/usr/local/zeebe/data/raft-partition/partitions/8/snapshots/51615149-1-1576111087120-126104534841528}, SnapshotImpl{position=124476742457264, path=/usr/local/zeebe/data/raft-partition/partitions/8/snapshots/50929172-1-1576110179037-124476742457264}, SnapshotImpl{position=122324971564288, path=/usr/local/zeebe/data/raft-partition/partitions/8/snapshots/50031228-1-1576109284278-122324971564288}]
 
I 2019-12-12T00:52:00.410749114Z 2019-12-12 00:52:00.410 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/8/runtime'.
 
I 2019-12-12T00:52:00.410789331Z 2019-12-12 00:52:00.410 [Broker-4-ZeebePartition-8] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Recovered state from snapshot 'SnapshotImpl{position=126104534841528, path=/usr/local/zeebe/data/raft-partition/partitions/8/snapshots/51615149-1-1576111087120-126104534841528}'
 
I 2019-12-12T00:52:00.424106780Z 2019-12-12 00:52:00.423 [partition-8-processor] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams - Recovering state of partition 8 from snapshot
 
I 2019-12-12T00:52:09.483754698Z 2019-12-12 00:52:09.482 [] [raft-server-4-raft-partition-partition-8] INFO  io.atomix.protocols.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-8} - AppendRequest{term=3, leader=4, prevLogIndex=52382943, prevLogTerm=1, entries=1, commitIndex=52382944} to 2 failed: java.util.concurrent.TimeoutException: Request type raft-partition-partition-8-append timed out in 5000 milliseconds
 
I 2019-12-12T00:52:10.502977907Z 2019-12-12 00:52:10.502 [partition-8-processor] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-1] INFO  io.zeebe.logstreams - Recovered state of partition 8 from snapshot at position 126104536860264
 
I 2019-12-12T00:52:10.504361261Z 2019-12-12 00:52:10.504 [partition-8-processor-snapshot-director] [zeebe-4.zeebe.long-running.svc.cluster.local:26501-zb-actors-2] DEBUG io.zeebe.logstreams.snapshot - The position of the last valid snapshot is '126104534841528'. Taking snapshots beyond this position.
 ```

</p>
</details>

**Environment:**
 - OS: [e.g. Linux]
 - Zeebe Version: [e.g. 0.20.0]
 - Configuration: [e.g. exporters etc.]

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

- splits destination folders for the different snapshot replications as they run in different thread context and both need to modify the same on-disk state
- Atomix snapshot replication will now write to raft-pending/ folder
- Zeebe snapshot replication will now write to pushed-pending/ folder

## Related issues

closes #3531, #4078, #3887, #3764

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Comment by npepinpe** [comment]
Will run a benchmark and report later.

File: broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java
Comment (deepthidevaki) [inline broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java]: We can use `FileUtil:ensureDirectoryExists` which you added in another PR to be consistent.

File: broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java
Comment (deepthidevaki) [inline broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java]: Doesn't purgeSnapshot does the cleanup?

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
:+1:

File: broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java
Comment (npepinpe) [reply broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java]: It now only does for the one pending directory it knows about - do you think it should also purge this one as well? I'm also not sure if it's a good idea for it do it anyway, because `purgeSnapshots` can be called from different thread contexts, so we might still run into the issue of deleting on going replications? Though it should be more fault tolerant

Comment (deepthidevaki) [reply broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java]: Sorry, I missed that DbSnapshot is shared between raft and atomix. Then it make sense to cleanup here. But then wouldn't we cleanup raft-pending snapshot while we purge snapshots here?

**Comment by npepinpe** [comment]
Blocked by https://github.com/zeebe-io/atomix/pull/205

File: broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java
Comment (npepinpe) [reply broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java]: As soon as the other PR is merged :+1:

File: broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/DbSnapshotStore.java
Comment (deepthidevaki) [inline broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/DbSnapshotStore.java]: This isn't used anymore. Can remove.

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
LGTM.

**Comment by npepinpe** [comment]
Thanks! Two things I addressed, and one question:

1. Removed the dead code
1. I moved the creation of the pending directory to the ZeebePartition because I generally dislike having constructors throw things like I/O exceptions or perform I/O, and I felt it's cleaner to fail the installation at that point by return a failed future - let me know what you think. It's not the cleanest now tbh, but I wasn't sure how else to structure it

And last, one question I thought, this is potentially "breaking" as we don't used the older `pending/` directory anymore, potentially leaving partial snapshots. Should we use `pending/` instead of `raft-pending/`? Long term, I would like to merge both replication mechanisms, so probably our only mechanism left would be the raft one, so we would go back to a single folder - might make sense to leave it named `pending/`. I thought it'd be more confusing initially, but now I think it's better :sweat_smile:

**Comment by npepinpe** [comment]
I think I broke something rebasing...don't review yet, I'll fix it :sweat_smile:

**Comment by npepinpe** [comment]
Also, regarding purging pending snapshots, we have a replication metric which I think may get out of sync. It's currently incremented/decremented entirely by the follower in the `ReplicationController` - increment when we start receiving a new snapshot, decrement if we invalidate it or complete it.

But let's say I start receiving snapshot X, then a new snapshot X + 1 is committed, and we purge X - but then I receive another chunk, so I increment the metric again. And maybe I never receive the last chunk, so I will never decrement it...

Would it be enough to decrement the metric here when we purge? Technically it may get incremented again by the replication controller, but then either it is eventually completed/invalidated or we purge it again and it's decremented.

**Comment by deepthidevaki** [comment]
> But let's say I start receiving snapshot X, then a new snapshot X + 1 is committed, and we purge X - but then I receive another chunk, so I increment the metric again. And maybe I never receive the last chunk, so I will never decrement it...

It won't increment the metric because it consider it as the continuation of ongoing replication of X. 

But I'm seeing another potential error here. start receiving snapshot X, then a new snapshot X + 1 is committed, and we purge pending X , received the last chunk of X , commit X because it thinks it has all chunks, but all the previous chunks were deleted :see_no_evil:  This scenario where X+1 commits before X is extremely rare though.

**Comment by npepinpe** [comment]
You're right though, that is potentially problematic...

Some possible solutions:

1. Only ensure that the directory exists when a new replication starts, which results in an error if you try to write a chunk after the purge
1. Calculate total snapshot checksum (e.g. checksum of every file so far, not just individual files), and on completion compute the total checksum again from the on-disk state and compare it

I would lean towards the second, but we could do this in another issue as this isn't a bug we're introducing here

**Review by deepthidevaki — APPROVED** [review]
:+1:

**Comment by npepinpe** [comment]
Build got killed :disappointed_relieved: 

bors retry

**Pull request merged**
